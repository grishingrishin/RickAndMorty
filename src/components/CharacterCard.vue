<script setup lang="ts">
import { computed } from 'vue';

import type { Character } from '@/libs/types';

type Props = {
	data: Character;
};

const props = withDefaults(defineProps<Props>(), {
	data: () => ({}) as Character,
});

// prettier-ignore
const allEpisodes = computed(() => {
	const {
		data: {
			episode
		}
	} = props;

	if (!Array.isArray(episode)) {
		return '';
	}

	const res = episode
		.map(item => item.replace(/\D/gi, ''))
		.join(',');

	return res;
});
</script>

<template>
	<div class="character-card">
		<div class="character-card__figure">
			<img :src="data.image" :alt="`Name: (${data.name})`" class="character-card__img" />
		</div>
		<div class="character-card__body">
			<div class="character-card__name">
				{{ data.name }}
			</div>
			<div class="character-card__status">
				{{ data.status }}
			</div>
			<template v-if="allEpisodes">
				<div class="character-card__episodes">Episodes: {{ allEpisodes }}</div>
			</template>
		</div>
	</div>
</template>

<style lang="scss">
.character-card {
	display: flex;
	flex-direction: column;
	gap: 14px;
	min-height: 320px;
	padding: 14px;

	&__body {
		display: flex;
		flex-direction: column;
		gap: 6px;
	}

	&__img {
		width: 100%;
		object-fit: cover;
		aspect-ratio: 1.1;
	}

	&__episodes {
		word-break: break-all;
	}
}
</style>
