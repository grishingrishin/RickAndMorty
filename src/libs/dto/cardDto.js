export function cardDto(data = {}) {
	return {
		picture: data.image ?? '',
		name: data.name ?? '',
		status: data.status ?? '',
		gender: data.gender ?? '',
		episodes: data.episode ?? [],
	};
}
