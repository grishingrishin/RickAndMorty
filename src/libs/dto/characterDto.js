export function characterDto(data = {}) {
	return {
		id: data.id ?? null,
		picture: data.image ?? '',
		name: data.name ?? '',
		status: data.status ?? '',
		gender: data.gender ?? '',
		episodes: data.episode ?? [],
	};
}
