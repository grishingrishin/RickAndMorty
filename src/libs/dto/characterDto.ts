import type { Character } from '@/libs/types';

// prettier-ignore
export function characterDto(
  data: Partial<Character> = {}
): Character {
  return {
    id: data.id ?? null,
    image: data.image ?? '',
    name: data.name ?? '',
    status: data.status ?? '',
    gender: data.gender ?? '',
    episode: data.episode ?? [],
  };
}
