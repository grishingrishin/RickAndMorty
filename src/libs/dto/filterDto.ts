import type { FilterParams } from '@/libs/types';

// prettier-ignore
export function filterDto(
  params: Partial<FilterParams> = {}
): FilterParams {
  return {
    status: params.status ?? '',
    gender: params.gender ?? '',
  };
}
