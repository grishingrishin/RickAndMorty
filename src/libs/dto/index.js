export * from '@/libs/dto/cardDto.js';
export * from '@/libs/dto/filterDto.js';
