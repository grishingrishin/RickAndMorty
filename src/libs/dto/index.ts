export * from '@/libs/dto/characterDto.ts';
export * from '@/libs/dto/filterDto.ts';
