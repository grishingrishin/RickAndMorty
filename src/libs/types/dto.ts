export type FilterParams = {
  status: string;
  gender: string;
};

export type Character = {
  id: number | null;
  image: string;
  name: string;
  status: string;
  gender: string;
  episode: string[];
};
