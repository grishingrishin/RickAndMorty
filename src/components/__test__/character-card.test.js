import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import { cardDto } from '@/libs/dto';

import CharacterCard from '@/components/CharacterCard.vue';

describe('The character card component', () => {
	test('Mounted component without episodes', () => {
		expect(CharacterCard).toBeTruthy();

		const wrapper = mount(CharacterCard, {
			props: {
				data: cardDto(),
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Mounted component with episodes', () => {
		expect(CharacterCard).toBeTruthy();

		const episode = [
			'https://rickandmortyapi.com/api/episode/1',
			'https://rickandmortyapi.com/api/episode/2',
			'https://rickandmortyapi.com/api/episode/3',
			'https://rickandmortyapi.com/api/episode/4',
			'https://rickandmortyapi.com/api/episode/5',
			'https://rickandmortyapi.com/api/episode/6',
			'https://rickandmortyapi.com/api/episode/7',
		];

		const wrapper = mount(CharacterCard, {
			props: {
				data: cardDto({
					episode,
				}),
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
