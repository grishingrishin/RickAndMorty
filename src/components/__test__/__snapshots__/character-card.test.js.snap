// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The character card component > Mounted component with episodes 1`] = `
"<div class="character-card">
  <div class="character-card__figure"><img src="" alt="Name: ()" class="character-card__img"></div>
  <div class="character-card__body">
    <div class="character-card__name"></div>
    <div class="character-card__status"></div>
    <div class="character-card__episodes">Episodes: 1,2,3,4,5,6,7</div>
  </div>
</div>"
`;

exports[`The character card component > Mounted component without episodes 1`] = `
"<div class="character-card">
  <div class="character-card__figure"><img src="" alt="Name: ()" class="character-card__img"></div>
  <div class="character-card__body">
    <div class="character-card__name"></div>
    <div class="character-card__status"></div>
    <!--v-if-->
  </div>
</div>"
`;
